### PR TITLE
fleetctl: destroy fix and more tests

### PR DESCRIPTION
**Pull request by tixxdz**
fleetctl destroy fix and more tests. It contains a follow up fix for the race condition that was discussed in this merged PR: https://github.com/coreos/fleet/pull/1417

### CODE REVIEW

**Comment by tixxdz** [comment]
functional tests pass!

File: fleetctl/unload_test.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
+// Copyright 2014 CoreOS, Inc.
```
Comment (jonboulle) [inline fleetctl/unload_test.go]: Date

File: fleetctl/unload_test.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	"time"
+
+	"github.com/coreos/fleet/client"
+	"github.com/coreos/fleet/job"
+	"github.com/coreos/fleet/machine"
+	"github.com/coreos/fleet/registry"
+	"github.com/coreos/fleet/unit"
+)
+
+type UnloadTestResults struct {
```
Comment (jonboulle) [inline fleetctl/unload_test.go]: General code hygiene: neither this type nor its members should be exported - make them all lower case.

File: fleetctl/unload_test.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	for _, r := range results {
+		var wg sync.WaitGroup
+		errchan := make(chan error)
+
+		cAPI = newFakeRegistryForUnload(unitPrefix, len(r.Units))
+
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			time.Sleep(2 * time.Microsecond)
```
Comment (jonboulle) [inline fleetctl/unload_test.go]: sleeps are bad - what are you trying to do?

File: fleetctl/unload_test.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	"github.com/coreos/fleet/unit"
+)
+
+type UnloadTestResults struct {
+	Description  string
+	Units        []string
+	ExpectedExit int
+}
+
+func newFakeRegistryForUnload(prefix string, unitCnt int) client.API {
```
Comment (jonboulle) [inline fleetctl/unload_test.go]: naming consistency: unitCnt -> unitCount - everywhere else you're spelling out words

File: fleetctl/unload_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+			[]string{"unload1", "unload2", "unload3", "unload4", "unload5"},
+			0,
+		},
+		{
+			"unload non-existent units",
+			[]string{"y1", "y2"},
+			0,
+		},
+		{
+			"attempt to unload available and non-available units",
```
Comment (jonboulle) [inline fleetctl/unload_test.go]: s/attempt to//
inconsistency between "non-available" and "non-existent" - just choose one

File: fleetctl/unload_test.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+	// We assume that we reached the desired state
+	for _, v := range real_units {
+		if job.JobState(v.DesiredState) != job.JobStateInactive {
+			errchan <- fmt.Errorf("Error: unit %s was not unloaded as requested", v.Name)
+		}
+	}
+}
+
+// TestRunUnloadUnits checks
```
Comment (jonboulle) [inline fleetctl/unload_test.go]: checks ...

File: fleetctl/unload_test.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	sharedFlags.NoBlock = true
+
+	exit := runUnloadUnit(r.Units)
+	if exit != r.ExpectedExit {
+		errchan <- fmt.Errorf("%s: expected exit code %d but received %d", r.Description, r.ExpectedExit, exit)
+	}
+
+	real_units, err := findUnits(r.Units)
+	if err != nil {
+		errchan <- fmt.Errorf("%v", err)
```
Comment (jonboulle) [inline fleetctl/unload_test.go]: errchan <- err ...

File: fleetctl/stop_test.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	"github.com/coreos/fleet/unit"
+)
+
+type StopTestResults struct {
+	Description  string
+	Units        []string
+	ExpectedExit int
+}
+
+func newFakeRegistryForStop(prefix string, unitCnt int) client.API {
```
Comment (jonboulle) [inline fleetctl/stop_test.go]: What's the difference between this and newFakeRegistryForUnload ? Do we need separate functions?

**Comment by jonboulle** [comment]
This looks fine at a high level but a few comments about the implementation. Looks like a fair amount of copy pasted code so hoping we can share that.

**Comment by tixxdz** [comment]
@jonboulle Updated and all comments handled, waiting to be green

File: fleetctl/fleetctl_test.go [hunk 9]
```diff
@@ -26,6 +28,68 @@ import (
 	"github.com/coreos/fleet/Godeps/_workspace/src/github.com/coreos/go-semver/semver"
 )
 
+type commandTestResults struct {
+	Description  string
```
Comment (jonboulle) [inline fleetctl/fleetctl_test.go]: these don't need to be exported

File: fleetctl/stop_test.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import (
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/coreos/fleet/job"
+)
+
+func doStopUnits(r commandTestResults, errchan chan error) {
+	sharedFlags.NoBlock = true
```
Comment (jonboulle) [inline fleetctl/stop_test.go]: Nit, this is a side effect. How about doing a sf := sharedFlags.NoBlock; sharedFlags.NoBlock = true; defer func() { sharedFlags.NoBlock = sf } within the context of the test loop?

**Comment by jonboulle** [comment]
Two more comments then LGTM

**Pull request merged**